File: scripts/day5-part2-simple.js
const fs = require("fs");

function solvePart2(input) {
  const ranges = input[0].split("\n").map((line) => {
    const [start, end] = line.split("-").map(Number);
    return { start, end };
  });

  // Merge overlapping ranges to avoid double counting
  const merged = [];
  const sorted = ranges.sort((a, b) => a.start - b.start);

  for (const range of sorted) {
    if (merged.length === 0) {
      merged.push({ ...range });
    } else {
      const last = merged[merged.length - 1];

      // If current range overlaps or is adjacent to the last one, merge them
      if (range.start <= last.end + 1) {
        last.end = Math.max(last.end, range.end);
      } else {
        merged.push({ ...range });
      }
    }
  }

  // Count all IDs in the merged ranges
  let totalCount = 0;
  for (const range of merged) {
    totalCount += range.end - range.start + 1;
  }

  return totalCount;
}

const input = fs
  .readFileSync("../inputs/day-5-input.txt", "utf-8")
  .trim()
  .split(/\r?\n\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 5 Part 2 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day5-part2-simple.py
import time

def solve_part2(input_data):
    parts = input_data.strip().split('\n\n')
    
    ranges = []
    for line in parts[0].split('\n'):
        start, end = map(int, line.split('-'))
        ranges.append((start, end))
    
    # Merge overlapping ranges to avoid double counting
    merged = []
    sorted_ranges = sorted(ranges, key=lambda x: x[0])
    
    for start, end in sorted_ranges:
        if not merged:
            merged.append([start, end])
        else:
            last = merged[-1]
            
            # If current range overlaps or is adjacent to the last one, merge them
            if start <= last[1] + 1:
                last[1] = max(last[1], end)
            else:
                merged.append([start, end])
    
    # Count all IDs in the merged ranges
    total_count = 0
    for start, end in merged:
        total_count += end - start + 1
    
    return total_count

# Read input
with open("../inputs/day-5-input.txt", "r") as f:
    input_data = f.read()

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 5 Part 2 - Simple)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
